pyoozie/tags.py

```diff
@@ -505,35 +634,69 @@ def __init__(
             job_tracker=None,             # type: typing.Optional[typing.Text]
             name_node=None,               # type: typing.Optional[typing.Text]
             job_xml_files=None,           # type: typing.Optional[JobXmlFilesType]
+            actions=None,                 # type: typing.Optional[_WorkflowEntity]
     ):
         # type: (...) -> None
         XMLSerializable.__init__(self, 'workflow-app')
-        self.name = validate_xml_name(name)
 
-        self.parameters = Parameters(parameters)
-        self.global_configuration = GlobalConfiguration(
+        # XML-document-related values
+        self.__name = validate_xml_name(name)
+        self.__parameters = Parameters(parameters)
+        self.__global_configuration = GlobalConfiguration(
             job_tracker=job_tracker,
             name_node=name_node,
             job_xml_files=job_xml_files,
             configuration=configuration
         )
-        self.credentials = credentials
+        self.__credentials = credentials or []
+        self.__actions = copy.deepcopy(actions)
+        self.__validate()
+
+    def __validate(self):  # type () -> None
+        action_identifiers = []
+        credentials_needed = set()
+
+        def _parse_action(action):
+            action_identifiers.append(action.identifier())
+            if hasattr(action, 'credential'):
```

Comment:
All entity identifiers need to be checked to find duplicates, but not all entities have credentials. I should probably rename action to entity.